### PR TITLE
Added Coords methods to convert to/from Astropy SkyCoord instances

### PR DESCRIPTION
**Pull request by ayshih**
For your consideration, this PR enables interacting back and forth with the Astropy coordinates framework (which is also used by SunPy).  It adds the following methods to `Coords`:

- `to_skycoord()` is a regular method which creates an Astropy `SkyCoord` instance from a `Coords` instance
- `from_skycoord()` is a class method which creates a `Coords` instance from an Astropy `SkyCoord` instance

The conversion utilizes the GEO coordinate frame, which is common to both libraries.

### CODE REVIEW

**Comment by drsteve** [comment]
Thanks for the PR @ayshih 

Tests are currently failing on this PR because of a broken URL for the NASA CDF library in our `.travis.yml` file (fixed in #288). If you can rebase against spacepy/master and push the missing commit then the Travis CI will be able to run.

While I'm supportive of interoperability between packages in the wider helio community, I'm also hesitant to approve this without a strategy for testing. Any thoughts? (Also paging @jtniehof for thoughts)

On the actual implementation:
- I'd probably not go via ISO time strings, for performance reasons, and would prefer to pass times using a numerical system that's shared between astropy and spacepy.
- There should also be appropriate checking for the presence of astropy so that spacepy returns a tailored error message with the `ImportError` if the dependency isn't present.
- The Earth radius used by IRBEM is present as an attribute (`Re`) on the `spacepy.coordinates.Coords` class, so there's no need to hard-code it here (although I note that it's not used anywhere, and incorrectly gives 6371000 m rather than 6371200 m). It's an instance attribute, not a class attribute, so won't be available until instantiation. That would either need a reorder of how the `Coords` instance is made/populated, or the `Re` attribute could be moved so it's a class attribute. I'd suggest having `Re` as a class method with the number corrected to 6371200.
- Single letter names in imports are strongly discouraged.

Note that it's on our plan to overhaul the coordinates module in SpacePy. Whenever that happens it should be transparent to users, but will provide higher accuracy conversions for the geophysical systems (see #52 ; also https://github.com/drsteve/LANLGeoMag/blob/master/libLanlGeoMag/Lgm_CTrans.c). Ideally I think we'd have an internal, well-tested conversion code (probably based on LANLGeoMag) and interoperability with other packages. That's still up for discussion though.

**Comment by ayshih** [comment]
> Tests are currently failing on this PR because of a broken URL for the NASA CDF library in our `.travis.yml` file (fixed in #288). If you can rebase against spacepy/master and push the missing commit then the Travis CI will be able to run.

Rebased (but I haven't yet added tests for these enhancements).

> While I'm supportive of interoperability between packages in the wider helio community, I'm also hesitant to approve this without a strategy for testing. Any thoughts? (Also paging @jtniehof for thoughts)

My suggestion is that the associated unit tests would be purely to check whether the Astropy SkyCoords are created properly and read in properly.  It seems overly complicated to do more than that (e.g., testing Astropy's transformations within your tests).

> * I'd probably not go via ISO time strings, for performance reasons, and would prefer to pass times using a numerical system that's shared between astropy and spacepy.

This was a conscious choice on my part because the ISO time string functionally retains a lot of precision.  Astropy Time supports many other scales (JD, MJD, TAI, etc.), but there may be precision loss when you go to extremes.

> * There should also be appropriate checking for the presence of astropy so that spacepy returns a tailored error message with the `ImportError` if the dependency isn't present.

Done (with a little bit of extra contortion now that `ModuleNotFoundError` exists in Python 3.6+)

> * The Earth radius used by IRBEM is present as an attribute (`Re`) on the `spacepy.coordinates.Coords` class, so there's no need to hard-code it here (although I note that it's not used anywhere, and incorrectly gives 6371000 m rather than 6371200 m). It's an instance attribute, not a class attribute, so won't be available until instantiation. That would either need a reorder of how the `Coords` instance is made/populated, or the `Re` attribute could be moved so it's a class attribute. I'd suggest having `Re` as a class attribute with the number corrected to 6371200.

Done (I wasn't sure whether the wrong value for `Re` was being used anywhere)

> * Single letter names in imports are strongly discouraged.

It's actually "standard" practice to use `u` for `astropy.units`, but I understand the resistance.  I've revised the code so I don't even need to explicitly import it.

> Note that it's on our plan to overhaul the coordinates module in SpacePy. Whenever that happens it should be transparent to users, but will provide higher accuracy conversions for the geophysical systems (see #52 ; also https://github.com/drsteve/LANLGeoMag/blob/master/libLanlGeoMag/Lgm_CTrans.c). Ideally I think we'd have an internal, well-tested conversion code (probably based on LANLGeoMag) and interoperability with other packages. That's still up for discussion though.

Great!  I've put a huge amount of effort into SunPy's coordinate framework, and I look forward to pairing it with a high-accuracy geophysical framework.

File: spacepy/coordinates.py [hunk 1]
```diff
@@ -391,3 +397,81 @@ def append(self, other):
...
+        =====
+        This method requires Astropy to be installed.
+
+        This method uses the GEO coordinate frame as the common frame between the two libraries.
+        '''
+        try:
+            from astropy.coordinates import SkyCoord
+            from astropy.time import Time
+        except ImportError as e:
+            raise e.__class__("This method requires Astropy to be installed.")
```
Comment (Cadair) [inline spacepy/coordinates.py]: It's good to propagate the error forward:
```suggestion
            raise e.__class__("This method requires Astropy to be installed.") from e
```

File: spacepy/coordinates.py [hunk 2]
```diff
@@ -391,3 +397,81 @@ def append(self, other):
...
+        Notes
+        =====
+        This method requires Astropy to be installed.
+
+        This method uses the GEO coordinate frame as the common frame between the two libraries.
+        '''
+        try:
+            from astropy.coordinates import SkyCoord
+        except ImportError as e:
+            raise e.__class__("This method requires Astropy to be installed.")
```
Comment (Cadair) [inline spacepy/coordinates.py]:
```suggestion
            raise e.__class__("This method requires Astropy to be installed.") from e
```

File: spacepy/coordinates.py (hunk 1 above)
Comment (ayshih) [reply spacepy/coordinates.py]: Done

File: spacepy/coordinates.py (hunk 2 above)
Comment (ayshih) [reply spacepy/coordinates.py]: Done

File: spacepy/coordinates.py (hunk 1 above)
Comment (ayshih) [reply spacepy/coordinates.py]: Reverted.  SpacePy still tests against Python 2.7, where this is not valid syntax.

File: spacepy/coordinates.py (hunk 2 above)
Comment (ayshih) [reply spacepy/coordinates.py]: Reverted.  SpacePy still tests against Python 2.7, where this is not valid syntax.

**Comment by ayshih** [comment]
@drsteve I've now added tests.  Any further comments?

In the future, when you overhaul the transformation code (using LGM or otherwise), I would strongly advocate that you take that opportunity to incorporate native support for `astropy.time` and `astropy.coordinates`, because that would make transformations frictionless across libraries.

**Comment by Cadair** [comment]
I would second that, it would be great if we had more common object classes across all of heliophysics.

File: spacepy/coordinates.py [hunk 3]
```diff
@@ -391,3 +397,78 @@ def append(self, other):
...
+        if self.ticks is None:
+            raise ValueError("This method requires the attribute `ticks` to be specified.")
+
+        try:
+            from astropy.coordinates import SkyCoord
+            from astropy.time import Time
+        except ImportError as e:
+            raise e.__class__("This method requires Astropy to be installed.")
+
+        # Convert to GEO (i.e., Astropy's ITRS)
```
Comment (jtniehof) [inline spacepy/coordinates.py:426]: This style of comment, statement, blank line soaks up vertical space and isn't adding much...it's obvoius it's converting to GEO, just need the ITRS info. Since the github multi-line suggestion thing doesn't work well, I'm envisioning something like:

```
coord = self.convert('GEO', 'car') # Astropy ITRS is GEO
data = coord.data * self.Re # metres
obstime = astropy.time.Time(coord.ticks.ISO, format='isot', scale='utc')
```

As @drsteve suggested, maybe use, say, TAI instead of ISO ticks?

File: spacepy/coordinates.py [hunk 4]
```diff
@@ -391,3 +397,78 @@ def append(self, other):
...
+        This method requires Astropy to be installed.
+
+        This method uses the GEO coordinate frame as the common frame between the two libraries.
+        '''
+        try:
+            from astropy.coordinates import SkyCoord
+        except ImportError as e:
+            raise e.__class__("This method requires Astropy to be installed.")
+
+        # Cast the input as a SkyCoord and then transform to Astropy's ITRS (i.e., GEO)
```
Comment (jtniehof) [inline spacepy/coordinates.py:465]: Similar to above

File: spacepy/coordinates.py [hunk 5]
```diff
@@ -391,3 +397,78 @@ def append(self, other):
...
+        skycoord = SkyCoord(skycoord).itrs
+
+        # Convert the SkyCoord's data to a Cartesian representation with units of R_earth
+        data = (skycoord.cartesian.xyz.to_value('m') / cls.Re).T
+
+        # Convert the SkyCoord's obstime to a Ticktock instance
+        ticks = spacepy.time.Ticktock(skycoord.obstime.utc.isot, 'ISO')
+
+        return Coords(data, 'GEO', 'car', ticks=ticks)
```
Comment (jtniehof) [inline spacepy/coordinates.py:474]:
```suggestion
        return cls(data, 'GEO', 'car', ticks=ticks)
```

File: tests/test_coordinates.py [hunk 6]
```diff
@@ -11,6 +11,10 @@
     import spacepy.irbempy as ib
 except ImportError:
     pass #tests will fail, but won't bring down the entire suite
+try:
```
Comment (jtniehof) [inline tests/test_coordinates.py:14]: I wonder about using a skip decorator on these, since astropy is optional. The reason for just forging ahead if irbem didn't import is because we still want to let other tests go ahead (to get that feedback in case there were other things wrong) but that's still a really pathalogical case, didn't seem worth skip-decorating the whole file. But for astropy coords, somebody might benefit from having a clean test run even without astropy as an optional dependency.

File: spacepy/coordinates.py (hunk 3 above)
Comment (drsteve) [reply spacepy/coordinates.py:426]: There's a _big_ nomenclature difference between TAI as used in key sections of the space physics community, and TAI as used by the broader astronomical and astrodynamics community (which is the correct one, of course). We use the "incorrect" version because it's fairly well entrenched in a number of key areas of magnetospheric physics. Our "TAI" is integer seconds since 1958/1/1, where the _proper_ TAI is offset from UTC by the number of leapseconds.
For a numeric time system, I'd probably use MJD. Although I think UTC datetimes get generated for pretty much any input time system, so it's probably least overhead to use those rather than forcing a conversion. If a conversion is desired, then MJD is my preference.
ISO strings are fine by me as long as we don't think the performance hit will be an issue.

TL;DR - MJD or UTC datetimes are my preference. ISO is fine if the performance hit is negligible.

File: spacepy/coordinates.py (hunk 5 above)
Comment (ayshih) [reply spacepy/coordinates.py]: Changed

File: tests/test_coordinates.py (hunk 6 above)
Comment (ayshih) [reply tests/test_coordinates.py:14]: Tests now decorated with `skipUnless`

File: spacepy/coordinates.py [hunk 7]
```diff
@@ -391,3 +397,78 @@ def append(self, other):
...
         return newobj
+
+    # -----------------------------------------------
+    def to_skycoord(self):
+        '''
+        Create an Astropy SkyCoord instance based on this instance
+
+        Returns
+        =======
+        out : Astropy SkyCoord instance
```
[the remark this answers is not in the document]
Comment (ayshih) [reply spacepy/coordinates.py]: Changed to intersphinx-style notation since you'll probably end up there eventually.

File: spacepy/coordinates.py [hunk 8]
```diff
@@ -391,3 +397,78 @@ def append(self, other):
...
+        =====
+        This method requires Astropy to be installed.
+
+        This method uses the GEO coordinate frame as the common frame between the two libraries.
+        '''
+        if self.ticks is None:
+            raise ValueError("This method requires the attribute `ticks` to be specified.")
+
+        try:
+            from astropy.coordinates import SkyCoord
```
[the remark this answers is not in the document]
Comment (ayshih) [reply spacepy/coordinates.py]: Changed here and elsewhere

File: spacepy/coordinates.py (hunk 4 above)
Comment (ayshih) [reply spacepy/coordinates.py]: Condensed

File: spacepy/coordinates.py (hunk 3 above)
Comment (ayshih) [reply spacepy/coordinates.py]: So...  It turns out that how you handle ISO, MJD, and UTC are all problematic because I believe they all repeat at the leap second.  After some experimentation, I believe the most sensible approach is to define a custom format for Astropy Time that matches your definition for "TAI".  That way, there is no degeneracy at the leap second and conversion overhead is kept to a minimum.

This custom time format is defined at the bottom of `spacepy.coordinates` to keep it conceptually grouped with the SkyCoord conversion functions, but of course it may make sense to move it to `spacepy.time`.

Example:
```python
>>> from spacepy.time import Ticktock
>>> Ticktock(1861920037.0, 'TAI').ISO
dmarray(['2017-01-01T00:00:00'], dtype='<U19')

>>> import spacepy.coordinates
>>> import astropy.time
>>> Time('2017-01-01 00:00:00', scale='utc').spacepy_tai
1861920037.0
>>> Time(1861920037.0, format='spacepy_tai').utc.iso
'2017-01-01 00:00:00.000'
```

**Comment by ayshih** [comment]
> Can you add Astropy, including the minimum version, to the soft dependencies in dependencies.rst?

Done

**Comment by jtniehof** [comment]
@drsteve and I looked over the time situation. Generally, as long as you're using a time system that can count leap seconds appropriately, Ticktock should be handling them properly. It's possible that some transforms might pass through intermediate representations that are losing leap seconds, in which case we'd like to identify and fix those.

Taking the leap second at the end of 2016, if I take an ISO timestamp for
the last non-leap second, it's two seconds to the beginning of 2017:
```
>>> tt = spacepy.time.Ticktock(['2016-12-31T23:59:59',
'2017-01-01T00:00:00'])
>>> tt[1] - tt[0]
[datetime.timedelta(0, 2)]
```

Contrasted with a year without a leap second:
```
>>> tt = spacepy.time.Ticktock(['2017-12-31T23:59:59',
'2018-01-01T00:00:00'])
>>> tt[1] - tt[0]
[datetime.timedelta(0, 1)]
```

It's true that this is a bit weird:
```
>>> spacepy.time.Ticktock('2016-12-31T23:59:59') +
datetime.timedelta(seconds=1)
Ticktock( [datetime.datetime(2017, 1, 1, 0, 0)], dtype=UTC)
```

But note the dtype is "UTC" and the SpacePy Ticktock handling of UTC is with datetimes, not really a true UTC. So we're looking into addition using objects other than timedeltas (#317).

If we hand in a TAI it does get handled properly:
```
>>> spacepy.time.Ticktock('2016-12-31T23:59:59').TAI[0]
1861920035.0
>>> tt = spacepy.time.Ticktock([1861920035, 1861920036, 1861920037],
dtype='TAI')
>>> tt.UTC
dmarray([datetime.datetime(2016, 12, 31, 23, 59, 59),
         datetime.datetime(2017, 1, 1, 0, 0),
         datetime.datetime(2017, 1, 1, 0, 0)], dtype=object)
```
UTC looks bad, but that's because datetime doesn't handle it. We do still
have the correct TAI:
```
>>> tt.TAI
dmarray([1861920035, 1861920036, 1861920037])
```

Now there does seem to be a problem with MJD:
```
>>> tt.MJD
dmarray([ 57753.99998843,  57754.        ,  57754.        ])
```

This is because our JD and MJD handling is touching Python datetime. So MJD is probably the right way to go but we're going to fix it first (#319)

The difference in "space physics" TAI is that it's just represented as a number of seconds rather than a timestamp, and as @drsteve said it's relative to the 1958-01-01 epoch.

I haven't dug deeply into AstroPy time, but working from the docs, looks like the easiest way to get a SpacePy TickTock into AstroPy is via GPS time, since AstroPy uses gps as a count of seconds in TAI time scale, just with a different reference epoch. So if, using our example above, you do:
```
>>> astropy_tt = astropy.time.Time(tt.TAI - 694656019.0, format='gps')
```

Then you can see it's converted properly:
```
>>> astropy_tt.format='isot'
>>> astropy_tt.utc
<Time object: scale='utc' format='isot' value=['2016-12-31T23:59:59.000'
'2016-12-31T23:59:60.000'
 '2017-01-01T00:00:00.000']>
```

Ticktock does do GPS time, but it does enough work through datetime that the conversion doesn't work properly here, on the list to fix (#320).

So medium-term: SpacePy will fix MJD and GPS time to eschew datetime and then one of those can be used. In the short-term, TAI plus an offset works. This should be a lot more straightforward than introducing a new subclass.

But we think it's better to separate out the time conversion into Ticktock, and should be able to get to that fairly directly (#321). Once that's done we'll loop back to this PR.

Thanks for bringing the time issues up. This was definitely a place that could use some polishing.

**Comment by jtniehof** [comment]
Since #417 is now merged, SpacePy's Ticktock supports representation as AstroPy time. So at a high level, this PR needs to be rebased to current master (conflict-resolved) and switched to using the Ticktock support. There are likely other small things to check but those are the biggies.

@ayshih , what's your preference for how to proceed? I can leave this all in your court, or I could do the rebase just to make sure the untangling goes well, or I could do the rebase and also the switchover of the time.

**Comment by ayshih** [comment]
Sure, I'll update this PR (although it might take me a little while to get to it).

A point of contention may be the minimum version of Astropy that will be supported, which you set in #417 as Astropy 1.0.  While your support plan is laudable, I think supporting 1.0, 1.1.1 (for Ubuntu 18.04 LTS), or even 2.0.4 (for Ubuntu 20.04 LTS) is risky for `astropy.coordinates`, and may be risky even for `astropy.time`.  Such ancient versions might work from the API perspective (i.e., errors are not raised), but there may be issues with the accuracy of calculations using those versions because of bugs that have been fixed in subsequent point releases (e.g., 2.0.16).

In this PR, I have currently listed Astropy 3.2 as the minimum version.  I know that Astropy 3.0+ may be problematic for SpacePy until you drop Python 2 support, but I'd recommend at least 2.0.x (preferably a point release more recent than 2.0.4).

**Comment by jtniehof** [comment]
As long as the code builds a reasonable time/coord object with AstroPy 1.0, I don't want to take the responsibility for whether the user might later take it into places that aren't suitable. Someone may genuinely get along fine with the earlier versions and no sense in blocking that.

**Comment by ehsteve** [comment]
@ayshih what are the plans on getting this finished up? Inquiring minds want to know ;)

**Comment by ayshih** [comment]
@ehsteve Soon™?

I'll see if I can give myself a solid kick on this next week.

**Comment by ayshih** [comment]
Okay, I *finally* got around to updating this PR.  Please review!

File: spacepy/coordinates.py (hunk 1 above)
Comment (Cadair) [reply spacepy/coordinates.py:422]: Does it still test against 2.7?

Comment (ayshih) [reply spacepy/coordinates.py:422]: As far as I can tell, SpacePy still tries to support 2.7

**Comment by drsteve** [comment]
Thanks for the update!

It looks like all of the test errors here are the same cause:
So that folks don't have to dig too deep in the GitHub Actions logs, here's one of the tracebacks
```
======================================================================
ERROR: test_to_skycoord_with_ticks (test_coordinates.coordsTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/runner/work/spacepy/spacepy/tests/test_coordinates.py", line 79, in test_to_skycoord_with_ticks
    assert np.all(sc.obstime.isclose(self.cvals.ticks.APT))
  File "/opt/hostedtoolcache/Python/3.9.6/x64/lib/python3.9/site-packages/astropy/time/core.py", line 1152, in __getattr__
    return self.__getattribute__(attr)
AttributeError: 'Time' object has no attribute 'isclose
```

The line `assert np.all(sc.obstime.isclose(self.cvals.ticks.APT))` in most of the new tests will fail on any astropy version prior to 4.2 (released on 2020-11-24), according to the astropy `CHANGELOG`.

For testing purposes, it'd probably make sense to do an explicit test of the times rather than using a new `astropy.time.Time` method. After all, the absence of the new `isclose` method won't affect whether the conversion worked or not. My suggestion would be to check that the _scale_, _format_, and _values_ are as expected.

**Comment by ayshih** [comment]
Ah, yes, I shouldn't have used `astropy.time.Time.isclose()`.  Rather than checking `scale`/`format`/`values` explicitly, I now check that the time difference is zero.

**Comment by ayshih** [comment]
> Based on the test failures and the astropy changelog, I think we're looking at requiring a bump in minimum supported version of astropy to `>=2.0` (or working around the `to_value` call in `from_skycoord`).

I think it'd be prudent to bump your minimum version of Astropy regardless, but I've replaced the `to_value()` call with something that should work on all versions of Astropy.

File: spacepy/coordinates.py [hunk 9]
```diff
@@ -459,7 +459,7 @@ def from_skycoord(cls, skycoord):
             raise e.__class__("This method requires Astropy to be installed.")
 
         skycoord = astropy.coordinates.SkyCoord(skycoord).itrs  # Astropy ITRS is GEO
-        data = (skycoord.cartesian.xyz.to_value('m') / cls.Re).T  # convert Cartesian to Re units
+        data = (skycoord.cartesian.xyz.to('m').value / cls.Re).T  # convert Cartesian to Re units
```
Comment (drsteve) [inline spacepy/coordinates.py:462]: Thanks for that change.

The CI cancelled the tests on 3.5 because they're running on an Ubuntu 16.04 image - don't worry about that. The failures look like they need this same change made on lines 75, 94, 109 & 125 of `test_coordinates.py`, but I'll let you double-check that...

Comment (ayshih) [reply spacepy/coordinates.py:462]: Oops!  Changes included now

**Review by drsteve — APPROVED** [review]

File: spacepy/coordinates.py [hunk 10]
```diff
@@ -392,6 +399,71 @@ def append(self, other):
...
+            raise ValueError("This method requires the attribute `ticks` to be specified.")
+
+        try:
+            import astropy.coordinates
+            import astropy.time
+        except ImportError as e:
+            raise e.__class__("This method requires Astropy to be installed.")
+
+        coord = self.convert('GEO', 'car')  # GEO is Astropy ITRS
+        data = coord.data * self.Re  # convert to meters
```
Comment (jtniehof) [inline spacepy/coordinates.py:428]: This assumes the input coordinates are in Re, but they can be (e.g.) km. We're not very good about handling this elsewhere--for instance, irbempy calls assume they're Re regardless of the units attribute--but it's probably worth doing something.

The fairly fancy approach would be to support Re, km, m and throw an error if they 1) aren't the same unit in all dimensions 2) not one of the above. I don't want to overburden this PR with fanciness. @drsteve , thoughts?

Comment (drsteve) [reply spacepy/coordinates.py:428]: My thought is that that's out of scope here. The documentation isn't as clear as it could be, but all Cartesian coordinate systems are assumed to be in Re, with spherical systems in degrees and Re. The only one that's different is geodetic, which is assumed/required to be in degrees and km. There's checking for that in #296, when we finally get to working through that (or breaking it for parts). The units are currently there to tell the user, not to let the user switch...

So this PR does the right thing for a correctly specified `Coords` per documentation. I'm happy as is.

**Review by jtniehof — APPROVED** [review]
Looks great, thanks for pushing through with this!

**Comment by ayshih** [comment]
Wonderful, thanks for the merge!
